File: app/src/main/java/com/pavelprymak/propodcast/presentation/paging/PodcastDataSourceFactory.java
package com.pavelprymak.propodcast.presentation.paging;

import androidx.annotation.NonNull;
import androidx.paging.DataSource;

import com.pavelprymak.propodcast.model.network.pojo.podcasts.PodcastItem;
import com.pavelprymak.propodcast.model.network.repo.PodcastRepoRx;
import com.pavelprymak.propodcast.presentation.common.PagingStateBatch;

import java.util.List;

public class PodcastDataSourceFactory extends DataSource.Factory<Integer, PodcastItem> {
    private final PagingStateBatch mPagingStateBatch;
    private final PodcastRepoRx mRepoRx;
    private final int mGenreId;
    private final String mRegion;
    private List<PodcastItem> mPrevLoadingList;

    public PodcastDataSourceFactory(PodcastRepoRx repoRx,
                                    PagingStateBatch pagingStateBatch,
                                    int genreId,
                                    String region) {
        mPagingStateBatch = pagingStateBatch;
        mRepoRx = repoRx;
        mGenreId = genreId;
        mRegion = region;
    }

    public PodcastDataSourceFactory(PodcastRepoRx repoRx,
                                    PagingStateBatch pagingStateBatch,
                                    int genreId,
                                    String region,
                                    List<PodcastItem> prevLoadingList) {
        mPagingStateBatch = pagingStateBatch;
        mRepoRx = repoRx;
        mGenreId = genreId;
        mRegion = region;
        mPrevLoadingList = prevLoadingList;
    }

    @NonNull
    @Override
    public DataSource<Integer, PodcastItem> create() {
        return new PodcastDataSource(mRepoRx, mPagingStateBatch, mGenreId, mRegion, mPrevLoadingList);
    }
}

File: app/src/main/java/com/pavelprymak/propodcast/presentation/paging/PodcastDataSourceFactory.kt
package com.pavelprymak.propodcast.presentation.paging

import androidx.paging.DataSource

import com.pavelprymak.propodcast.model.network.pojo.podcasts.PodcastItem
import com.pavelprymak.propodcast.model.network.repo.PodcastRepoRx
import com.pavelprymak.propodcast.presentation.common.PagingStateBatch

class PodcastDataSourceFactory : DataSource.Factory<Int, PodcastItem> {
    private val mPagingStateBatch: PagingStateBatch
    private val mRepoRx: PodcastRepoRx
    private val mGenreId: Int
    private val mRegion: String
    private var mPrevLoadingList: List<PodcastItem>? = null

    constructor(
        repoRx: PodcastRepoRx,
        pagingStateBatch: PagingStateBatch,
        genreId: Int,
        region: String
    ) {
        mPagingStateBatch = pagingStateBatch
        mRepoRx = repoRx
        mGenreId = genreId
        mRegion = region
    }

    constructor(
        repoRx: PodcastRepoRx,
        pagingStateBatch: PagingStateBatch,
        genreId: Int,
        region: String,
        prevLoadingList: List<PodcastItem>?
    ) {
        mPagingStateBatch = pagingStateBatch
        mRepoRx = repoRx
        mGenreId = genreId
        mRegion = region
        mPrevLoadingList = prevLoadingList
    }

    override fun create(): DataSource<Int, PodcastItem> {
        return PodcastDataSource(mRepoRx, mPagingStateBatch, mGenreId, mRegion, mPrevLoadingList)
    }
}
